### PR TITLE
[Issue #1745] Setup transformation process structure & transform an opportunity

### LINKED ISSUE #1745: [Task]: Setup the transformation process script
### Summary

Just setup the script and transform a single table:
https://app.gitbook.com/o/cFcvhi6d0nlLyH2VzVgn/s/v1V0jIH7mb7Yb3jlNrgk/engineering/learnings/opportunity-endpoint-data-model#elt-transformations

### Acceptance criteria

- [ ] Setup the script with basic logging / metrics
- [ ] Transforms just the opportunity table (other tables in later tickets)

### PR DESCRIPTION
**Pull request by chouinar**
## Summary
Fixes #1745

## Changes proposed
Setup the transformation process script structure

Implement some shared utilities that subsequent PRs will use

Implement the transformation logic for opportunities

## Context for reviewers
A lot of setup in this PR, a lot that can be reused in the subsequent PRs to add transformations for the other sets of tables. Tried to make sure those would require refactoring or pulling out implementation details by setting up utils like the timestamp conversions + initial query to the DB to fetch the transforming records.

As far as the implementation goes, determining what needs to be transformed is pretty simple - the `transformed_at` column is null. There is then a second column that says whether the record should be deleted or not. When we query the staging tables, we also join with the relevant table we'll transform to (opportunity in this case), that way we already have the "source" and "destination" records and just need to modify the destination record (or create it if its an insert).

### CODE REVIEW

File: api/src/data_migration/transformation/transform_oracle_data.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+S = TypeVar("S", bound=Base)
+D = TypeVar("D", bound=Base)
+
+from src.constants.lookup_constants import OpportunityCategory
+
+logger = logging.getLogger(__name__)
+
+
+class TransformOracleData(Task):
```
Comment (chouinar) [inline api/src/data_migration/transformation/transform_oracle_data.py:21]: @jamesbursa - FYI I'm imagine that the script we'll have for the copy + transform will just look like:

```
def whatever_entrypoint_func(db_session):
     ExtractLoadOracleData(db_session).run()
     TransformOracleData(db_session).run()
```

**Review by jamesbursa — DISMISSED** [review]
Looks good, thanks

File: api/src/data_migration/transformation/transform_oracle_data_task.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        logger.info("Processing opportunity", extra=extra)
+
+        if source_opportunity.is_deleted:
+            logger.info("Deleting opportunity", extra=extra)
+
+            if target_opportunity is None:
+                raise ValueError("Cannot delete opportunity as it does not exist")
+
+            self.increment(self.Metrics.TOTAL_RECORDS_DELETED)
+            self.db_session.delete(target_opportunity)
```
[the remark this answers is not in the document]
Comment (chouinar) [reply api/src/data_migration/transformation/transform_oracle_data_task.py:104]: We have the cascades setup on the relationships, so its safe to delete with the foreign keys.

File: api/src/util/datetime_util.py [hunk 3]
```diff
@@ -31,6 +31,11 @@ def adjust_timezone(timestamp: datetime, timezone_str: str) -> datetime:
     return timestamp.astimezone(new_timezone)
 
 
+def make_timezone_aware(timestamp: datetime, timezone_str: str) -> datetime:
+    new_timezone = pytz.timezone(timezone_str)
+    return new_timezone.localize(timestamp)
```
[the remark this answers is not in the document]
Comment (chouinar) [reply api/src/util/datetime_util.py:39]: That seems to handle a weird pytz case I was seeing where using `replace` with its timezones was adding 4 minutes - will adjust.

File: api/src/data_migration/transformation/transform_oracle_data_task.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    # Convert the source timestamps to UTC
+    # Note: the type ignores are because created_date/last_upd_date are added
+    #       on the individual class definitions, not the base class - due to how
+    #       we need to maintain the column order of the legacy system.
+    #       Every legacy table does have these columns.
+    created_timestamp = convert_est_timestamp_to_utc(source.created_date)  # type: ignore[attr-defined]
+    updated_timestamp = convert_est_timestamp_to_utc(source.last_upd_date)  # type: ignore[attr-defined]
+
+    if created_timestamp is not None:
+        target.created_at = created_timestamp
```
[the remark this answers is not in the document]
Comment (chouinar) [reply api/src/data_migration/transformation/transform_oracle_data_task.py:217]: How would you suggest we set this up and use it? I set them this way so when someone fetches them from our API, they have the same values as the legacy system.

How would we want to map these for the API if we add two more columns? A user shouldn't need to be aware that we technically updated everything when we imported it (that's irrelevant to them).

File: api/src/data_migration/transformation/transform_oracle_data_task.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        # confusion in the parsing so it ends up assuming everything is Any
+        # So just cast it to a simpler type that doesn't confuse anything
+        return cast(
+            list[Tuple[S, D | None]],
+            self.db_session.execute(
+                select(source_model, destination_model)
+                .join(destination_model, *join_clause, isouter=True)
+                .where(source_model.transformed_at.is_(None))
+                .execution_options(yield_per=5000)
+            ).all(),
```
[the remark this answers is not in the document]
Comment (chouinar) [reply api/src/data_migration/transformation/transform_oracle_data_task.py:68]: At this point I just implicitly add it to everything even though I don't think it does anything meaningful in SQLAlchemy 2.0. I do know that yield_per changed a bit in SQLAlchemy 2.0 - so it is always streamed in batches implicitly (this is very obvious if you run the batch with a lot of records as it'll "pause" briefly on each group). https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#fetching-large-result-sets-with-yield-per

**Review by jamesbursa — APPROVED** [review]

**Pull request merged**